Add conditional for CPT archives and CPT term archives

mikeyarce opened the pull request:
## Description
When you register a custom post type with a taxonomy, the archive pages don't have the right metadata, they pick up the metadata from the first item in the loop instead of the main archive page.

What I'm trying here is adding the `is_post_type_archive()` conditional and switching to getting the queried object instead of just calling `get_the_category()` as that doesn't work in all cases.

Edit:  I found that there's another case that I wanted to target: custom post type taxonomy archives.  I've addressed that and added unit tests for that as well. 

Fixes #327 

## Motivation and Context
Fixing this for custom post type archives

## How Has This Been Tested?
1. On your test site, make sure you have a CPT with a category
2. Also make sure you have a regular post type with a category
3. Check this PR out to your test site
4. Go to the Archive pages for both the CPT and the regular post category 
5. Check to see if the right metadata is present for them both.


## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)


Question:  Should we also check if the post type is currently added to the Parse.ly options?

mikeyarce committed:
Add conditional for cpt archives

Change to get queried object instead of category.

GaryJones commented:
This could do with some tests.

mikeyarce committed:
Change conditional to is_archive

`is_archive()` is better in this case because it also targets taxonomy archives as well as custom post type archives.

mikeyarce committed:
Add unit tests for CPT archive metadata

mikeyarce committed:
Add unit test for CPT term archive

mikeyarce committed:
Beautify code

GaryJones commented on tests/structured-data/class-archive-post-test.php:
```diff
@@ -233,6 +233,143 @@ public function test_term_archive() {
+		$parsely_options = get_option( \Parsely::OPTIONS_KEY );
+
+		// Register custom post type and custom taxonomy for that post type.
+		register_post_type( 'custom_post_type', array( 'public' => true ) );
```
Is the fact that that custom taxonomy is associated with a CPT, relevant here, or just a distraction?

Could you have created a CTax and associated it with a `post`, and been able to test the same thing?

mikeyarce replied:
Good question!   I was testing this way because a bug report came in with this specific scenario, but it's really the same regardless of Custom Post Type or not. 

I think for simplicity I'll remove the CPT out of the equation here.

GaryJones commented on tests/structured-data/class-archive-post-test.php:
```diff
@@ -233,6 +233,143 @@ public function test_term_archive() {
+			'custom_post_type',
+			array(
+				'public'      => true,
+				'has_archive' => 'cpt-tax-archive',
```
If you've got a CPT, why give it an archive of `cpt-tax-archive`? The `-tax` bit seems irrelevant since there's no Custom Taxonomy aspect of this test.

GaryJones commented on tests/structured-data/class-archive-post-test.php:
```diff
@@ -233,6 +233,143 @@ public function test_term_archive() {
+		// The query should be for a taxonomy archive.
+		self::assertQueryTrue( 'is_archive', 'is_tax' );
+
+		// Create the structured data for that category.
+		// The category metadata doesn't use the post data, but the construction method requires it for now.
```
```suggestion
		// Create the structured data for that term archive.
		// The term archive metadata doesn't use the post data, but the construction method requires it for now.
```

GaryJones commented on tests/structured-data/class-archive-post-test.php:
```diff
@@ -233,6 +233,143 @@ public function test_term_archive() {
+		// Check the required properties exist.
+		$this->assert_data_has_required_properties( $structured_data );
+
+		// The headline should be the category name.
```
```suggestion
		// The headline should be the term name.
```

GaryJones commented on tests/structured-data/class-archive-post-test.php:
```diff
@@ -233,6 +233,143 @@ public function test_term_archive() {
+			array(
+				'taxonomy' => 'custom_tax',
+				'slug'     => 'term',
+				'name'     => 'Custom Taxonomy',
```
```suggestion
				'name'     => 'Custom Taxonomy Term',
```

It's easy to get confused between the name of the taxonomy, and the name of a term in that taxonomy, so keep the distinction clear.

The assertion on this name will also need updating.

GaryJones commented on src/class-parsely.php:
```diff
@@ -885,9 +885,8 @@ public function construct_parsely_metadata( array $parsely_options, $post ) {
-		} elseif ( is_category() ) {
-			$category                 = get_the_category();
-			$category                 = $category[0];
+		} elseif ( is_category() || is_archive() ) {
```
`is_archive()` is true for many types of archives, like date or tag archives, which are below this conditional, and so requests to a date archive, for instance, will be captured here, instead of below, which will change the behaviour; there may not have been unit tests yet to flag this.

For now, use `is_post_type_archive()` as the most narrowest scope conditional for what you're trying to address here.

mikeyarce replied:
`is_post_type_archive()` will only be true for the custom post type archive, so what if we did:

`(  is_category() || is_post_type_archive() || is_tax() ) `

GaryJones replied:
Yep, that should work 👍🏻

mikeyarce committed:
Remove custom post type to simplify test

It doesn’t change the functionality to have a custom post type or not, so to simplify we’ll remove that from the test.

mikeyarce committed:
Change cpt-tax-archive to cpt-archive

mikeyarce committed:
Fix code comments

mikeyarce committed:
Clarify the name of the custom taxonomy

GaryJones commented on tests/structured-data/class-archive-post-test.php:
```diff
@@ -318,7 +318,7 @@ public function test_archive_custom_post_type() {
 	 * @group metadata
 	 */
-	public function test_term_archive_custom_post_type() {
+	public function test_custom_term_archive() {
```
```suggestion
	public function test_custom_taxonomy_term_archive() {
```

All _terms_ are custom ;-) - it's the _taxonomy_ that's custom that is the focus here.

GaryJones reviewed (changes requested):
Fix the condition as you suggested, tweak the test method name as I've now added, and then this is good to land. Thank you Mikey 👍🏻

mikeyarce committed:
Update check for cpt and taxonomy archives

mikeyarce committed:
Change test name

GaryJones reviewed (approved):
:shipit:

pauarge committed:
Merge branch 'develop' into fix/cpt-categories

pauarge commented:
I have merged `develop` into this branch. I kept the changes, but I had to split the tests into two separate files to follow the pattern that we have adopted since we did this PR.

pauarge committed:
Fix linting issues

pauarge committed:
Fix linting issues

pauarge reviewed (approved):
LGTM

GaryJones commented on tests/StructuredData/CustomPostTypeArchiveTest.php:
```diff
@@ -0,0 +1,83 @@
+		// Check the required properties exist.
+		$this->assert_data_has_required_properties( $structured_data );
+
+		// The headline should be the category name.
```
```suggestion
		// The headline should be the CPT name.
```

GaryJones commented on tests/StructuredData/CustomPostTypeArchiveTest.php:
```diff
@@ -0,0 +1,83 @@
+		// The query should be for a custom post type archive.
+		self::assertQueryTrue( 'is_archive', 'is_post_type_archive' );
+
+		// Create the structured data for that category.
```
```suggestion
		// Create the structured data for that CPT.
```

GaryJones commented on tests/StructuredData/CustomPostTypeArchiveTest.php:
```diff
@@ -0,0 +1,83 @@
+		self::assertQueryTrue( 'is_archive', 'is_post_type_archive' );
+
+		// Create the structured data for that category.
+		// The category metadata doesn't use the post data, but the construction method requires it for now.
```
```suggestion
		// The CPT archive metadata doesn't use the post data, but the construction method requires it for now.
```

GaryJones commented on tests/StructuredData/CustomTaxonomyTermArchiveTest.php:
```diff
@@ -0,0 +1,83 @@
+
+		$term_link = get_term_link( $term );
+
+		// Flush rewrite rules after creating new post type with archive.
```
```suggestion
		// Flush rewrite rules after creating new taxonomy type.
```

GaryJones commented on tests/StructuredData/CustomPostTypeArchiveTest.php:
```diff
@@ -0,0 +1,83 @@
+	 * @uses \Parsely::update_metadata_endpoint
+	 * @group metadata
+	 */
+	public function test_archive_custom_post_type() {
```
```suggestion
	public function test_metadata_is_correctly_constructed_for_custom_post_type_archive() {
```

GaryJones commented on tests/StructuredData/CustomTaxonomyTermArchiveTest.php:
```diff
@@ -0,0 +1,83 @@
+	 * @uses \Parsely::update_metadata_endpoint
+	 * @group metadata
+	 */
+	public function test_custom_taxonomy_term_archive() {
```
```suggestion
	public function test_metadata_is_correctly_constructed_for_custom_taxonomy_term_archive() {
```

GaryJones reviewed (changes requested):
Minor tweaks.

pauarge committed:
Apply suggestions from code review

Co-authored-by: Gary Jones <gary.jones@automattic.com>

pauarge committed:
Merge branch 'develop' into fix/cpt-categories

pauarge committed:
Merge branch 'develop' into fix/cpt-categories
